### ISSUE TITLE
Possible bug

### ISSUE THREAD

**Issue by trangham283** [rtmdrr/testSignificanceNLP#3]
In this [line](https://github.com/rtmdrr/testSignificanceNLP/blob/08fc2dbf4dc7d1b89d5830081d13baa199cdb22d/testSignificance.py#L100), shouldn't the sign be flipped? i.e. increment r if delta > 2*orig_delta. 
That's according to the Berg-Kirkpatrick et al. (2012) paper, unless I'm misunderstanding something. 
Thanks!

**Comment by texttheater** [comment]
So... this was fixed for Bootstrap, but not for Permutation?

**Comment by texttheater** [comment]
Someone correct me please, but I think the original version is correct assuming that data_A is the output of the baseline and data_B is the output of the algorithm to be tested. If you swap the two, the sign of the deltas is going to be flipped, so you have to flip the comparator.

**Comment by mgrenander** [comment]
For anyone looking to use this repository, I want to add my understanding that the original version does not have a bug, and #4 actually introduces a bug. 

I believe that Berg-Kirkpatrick et al. (2012)'s description of the bootstrap in Figure 1 contains an error in step 4. The intuition is that $s$ is counting the "success" condition, where A beats B by margin $2\delta$. If $s$ is very large, then the p-value should be small, so that the null hypothesis can then be rejected. So in step 4, the p-value should not be computed as $\frac{s}{b}$, but rather as $1 - \frac{s}{b}$.

The original implementation in this repository is therefore correct because they are not counting $\delta(x^{(i)}) > 2\delta(x)$, but the rather the *complement*, i.e. by flipping the comparator. In this case, it is correct to estimate the p-value as $\frac{s}{b}$.

tl;dr Use the original implementation and not the version after #4 was merged.

**Comment by trangham283** [comment]
Hmmm my understanding is that you're not counting successes in `s`, it is tallying how often you'd still observe A beating B _pairwise_ by such a large margin as 2\delta under H0, when the results in both sets are shuffled and you would no longer expect such wins. 

As a concrete example, here I created a test dataset where A clearly beats B:

```python
import numpy as np

# Version 1 (original)
def bootstrap_v1(data_A, data_B, n, R):
    delta_orig = float(sum([x - y for x, y in zip(data_A, data_B)])) / n
    r = 0
    for x in range(0, R):
        temp_A = []
        temp_B = []
        samples = np.random.randint(0,n,n)
        for samp in samples:
            temp_A.append(data_A[samp])
            temp_B.append(data_B[samp])
        delta = float(sum([x - y for x, y in zip(temp_A, temp_B)])) / n
        if (delta < 2*delta_orig):
            r = r + 1
    pval = float(r)/(R)
    return pval

# Version 2 (after PR #4)
def bootstrap_v2(data_A, data_B, n, R):
    delta_orig = float(sum([x - y for x, y in zip(data_A, data_B)])) / n
    r = 0
    for x in range(0, R):
        temp_A = []
        temp_B = []
        samples = np.random.randint(0,n,n)
        for samp in samples:
            temp_A.append(data_A[samp])
            temp_B.append(data_B[samp])
        delta = float(sum([x - y for x, y in zip(temp_A, temp_B)])) / n
        if (delta > 2*delta_orig):
            r = r + 1
    pval = float(r)/(R)
    return pval

# Example
data_A = [90, 92, 85, 88, 91, 87, 93, 89, 94, 88]
data_B = [85, 88, 82, 85, 87, 84, 89, 86, 90, 85]

n = len(data_A)
R = 1000

print(f"Version 1 p-val: {bootstrap_v1(data_A, data_B, n, R)}")
print(f"Version 2 p-val: {bootstrap_v2(data_A, data_B, n, R)}")

```

Then version 1 gives `p = 1.0` while version 2 gives `p = 0.0`, which is what I would expect.
